wstr/src/ptr.rs

```diff
@@ -47,19 +45,54 @@ pub struct WStr {
     _repr: [()],
 }
 
-impl ToOwned for WStr {
-    type Owned = WString;
+/// The metadata of a `WStr` pointer. This is always 4 bytes wide, even on 64-bits targets.
+///
+/// The layout of `WStr` depends on the value of `self.is_wide()`:
+///  - if `false`, it has the layout of `[u8; self.len()]`;
+///  - if `true`, it has the layout of `[u16; self.len()]`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct WStrMetadata(u32);
+
+impl WStrMetadata {
+    /// SAFETY: raw must fit in a u32
+    #[inline(always)]
+    const unsafe fn from_usize(raw: usize) -> Self {
+        if raw > u32::MAX as usize {
+            if cfg!(debug_assertions) {
+                panic!("invalid WStr metadata");
+            } else {
+                core::hint::unreachable_unchecked()
+            }
+        }
```

Comment:
Could this just be `debug_assert!(raw <= u32::MAX as usize)`?

Reply:
The `unreachable_unchecked` is there to (at least in theory) give more information to the optimizer and make everything *Faster*™.

I haven't looked at the exact assembly, but removing the `unreachable_unchecked` call makes a release build of `ruffle_desktop` roughly 3kB bigger, so clearly some branches are getting optimized out.